### Анализ данных посредством SQL запросов 
Наша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные должны сформулировать ценностное предложение для нового продукта.

#### Часть 1. Изучение общей информаци
- Загрузка библиотек, подключение к БД, получение общей информации, изученеие файлов с данными.

#### Часть 1. Исследование

#### Общий вывод

### Часть 1. Изучение общей информаци

In [1]:
import pandas as pd
from sqlalchemy import create_engine 

In [2]:
#pip install psycopg2 

In [3]:
# устанавливаем параметры
db_config = {'user': '******', # имя пользователя
 'pwd': '*******', # пароль
 'host': '******',
 'port': 6432, # порт подключения
 'db': '******'} # название базы данных

In [4]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db']) 

In [5]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [6]:
def general_information(data):
    query = data
    data = pd.io.sql.read_sql(query, con = engine)  # sql-запрос
    data.info()
    print()
    display(data.head())
    print()
    print('Дубликатов', data.duplicated().sum(), 'или', round(data.duplicated().sum() / len(data) * 100), '%')

In [7]:
general_information(''' SELECT * FROM books''')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB



book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


Дубликатов 0 или 0 %


**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [8]:
general_information(''' SELECT * FROM authors''')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB



author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


Дубликатов 0 или 0 %


**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [9]:
general_information(''' SELECT * FROM publishers''')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB



publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


Дубликатов 0 или 0 %


**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [10]:
general_information(''' SELECT * FROM ratings''')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB



rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


Дубликатов 0 или 0 %


**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [11]:
general_information(''' SELECT * FROM reviews''')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB



review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


Дубликатов 0 или 0 %


**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

#### Вывод
- Иммеются таблицы с данными:
  - Таблица `books`. Содержит данные о книгах:`book_id` — идентификатор книги, `author_id` — идентификатор автора, `title` — название книги, `num_pages` — количество страниц, `publication_date` — дата публикации книги, `publisher_id` — идентификатор издателя.
  - Таблица `authors`. Содержит данные об авторах: `author_id` — идентификатор автора, `author` — имя автора.
  - Таблица `publishers`. Содержит данные об издательствах: `publisher_id` — идентификатор издательства, `publisher` — название издательства
  - Таблица `ratings`. Содержит данные о пользовательских оценках книг: `rating_id` — идентификатор оценки, `book_id` — идентификатор книги, `username` — имя пользователя, оставившего оценку, `rating` — оценка книги.
  - Таблица `reviews`. Содержит данные о пользовательских обзорах на книги: `review_id` — идентификатор обзора, `book_id` — идентификатор книги, `username` — имя пользователя, написавшего обзор, `text` — текст обзора.
  
  Таблицы связанны между собой с помощью ключей и имеют связи один ко многим, это ключ "book_id" и один к одному, это ключи "author_id" и publisher_id 
  
- Первый взгляд на таблицу. Можно заменить тип данных в соответствии с типом данных представленных в столбце дат - в формат даты. Но делать не будем, т.к. пока нам это не принципиально.
- Пропусков и дубликатов не обнаружено.

### Часть 1. Исследование

In [12]:
# функция для выполнения запосов
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

#### Посчитаем, сколько книг вышло после 1 января 2000 года;

In [13]:
sql = '''SELECT COUNT(*) FROM books
         WHERE publication_date > '01.01.2000'
      '''

In [14]:
select(sql)

count
0    819

После 1 января 2000 года вышло 819 книг

#### Для каждой книги посчитаем количество обзоров и среднюю оценку

Сджойним с левым соединением, т.к. вполне возможно, что не у всех книг есть оценки и обзоры

In [15]:
sql = '''SELECT title,
                COUNT(title),
                AVG(rating)
         FROM books LEFT JOIN  ratings ON books.book_id = ratings.book_id
                    LEFT JOIN reviews ON books.book_id = reviews.book_id
         GROUP BY title
         ORDER BY AVG(rating) desc
      '''

In [16]:
select(sql)

title  count   avg
0    Captivating: Unveiling the Mystery of a Woman'...      4  5.00
1                                        Evening Class      4  5.00
2    In the Hand of the Goddess (Song of the Liones...      6  5.00
3                   The Big Bad Wolf  (Alex Cross  #9)      4  5.00
4                        A Dirty Job (Grim Reaper  #1)     16  5.00
..                                                 ...    ...   ...
994  The World Is Flat: A Brief History of the Twen...     12  2.25
995                  His Excellency: George Washington      4  2.00
996                                              Junky      4  2.00
997                                      Drowning Ruth      9  2.00
998                               Harvesting the Heart      4  1.50

[999 rows x 3 columns]

Куда-то потерялась одна книжка) было 1000 строк в таблице books

In [17]:
sql = '''SELECT books.book_id,
                title,
                COUNT(books.book_id),
                AVG(rating)
         FROM books LEFT JOIN  ratings ON books.book_id = ratings.book_id
                    LEFT JOIN reviews ON books.book_id = reviews.book_id
         GROUP BY books.book_id
         ORDER BY AVG(rating) desc
      '''

In [18]:
select(sql)

book_id                                              title  count   avg
0        518               Pop Goes the Weasel (Alex Cross  #5)      4  5.00
1        732  The Ghost Map: The Story of London's Most Terr...      4  5.00
2        347  In the Hand of the Goddess (Song of the Liones...      6  5.00
3        610                           Tai-Pan (Asian Saga  #2)      4  5.00
4        330  How to Be a Domestic Goddess: Baking and the A...      2  5.00
..       ...                                                ...    ...   ...
995      915  The World Is Flat: A Brief History of the Twen...     12  2.25
996      371                                              Junky      4  2.00
997      316                  His Excellency: George Washington      4  2.00
998      202                                      Drowning Ruth      9  2.00
999      303                               Harvesting the Heart      4  1.50

[1000 rows x 4 columns]

Нашлася, видимо есть повторяющееся название книги, т.к. какое-то значение у этой книги отличается. Удивительно, что только одна такая ктнига, ведь это м.б и другое количество страниц и дата публикации и издатель, при чем теоретически у каждой книги нашей таблицы, а практически - у самых популярных.

#### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключили из анализа брошюры

In [20]:
sql = '''  SELECT  publisher 
           FROM publishers
           WHERE publisher_id = (SELECT publishers.publisher_id
                                 FROM (SELECT publisher_id,
                                              count(num_pages)
                                       FROM books 
                                       GROUP BY publisher_id
                                       HAVING publisher_id IN (SELECT publisher_id
                                                               FROM books
                                                               WHERE num_pages > 50)
                                       ORDER BY count(num_pages) desc
                                       LIMIT 1) query_in_1
                                 INNER JOIN publishers ON publishers.publisher_id = query_in_1.publisher_id)
         
      '''

In [21]:
select(sql)

publisher
0  Penguin Books

После смерти Аллена Лейна (основателя) в 1970 год издательство оставило в качестве главного направления издание книг в бумажных обложках. Если в бумажных, а не в кортонных, значит дешевле, значит и пользуется спросом, т.к. спрос рождает предложение. В нашем случае самое большое выпущенное количество книг.

#### Определим автора с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками

In [22]:
sql = ''' SELECT author 
          FROM authors
          WHERE author_id = (SELECT  author_id              
                             FROM books
                             WHERE book_id = (SELECT books.book_id                 
                                              FROM (SELECT book_id, AVG(rating)
                                                    FROM ratings
                                                    GROUP BY book_id
                                                    HAVING book_id IN (SELECT book_id
                                                                       FROM ratings
                                                                       GROUP BY book_id
                                                                       HAVING COUNT(rating) >= 50)
                                                    ORDER BY AVG(rating) desc
                                                    LIMIT 1) query_in_1
                                               INNER JOIN books ON books.book_id = query_in_1.book_id))                        
      '''
         

In [23]:
select(sql)

author
0  J.K. Rowling/Mary GrandPré

Джоан Роулинг — британская писательница, сценаристка и кинопродюсер, Мэри Грандпре — американский иллюстратор и писатель, эти два человека наиболее известные своим авторством и своими иллюстрациями изданий книг о Гарри Поттере. Понятно теперь, почему именно они победили в этом рейтинге)

#### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [24]:
sql = ''' SELECT ROUND(SUM(query_in_1.cnt) / COUNT(query_in_1.cnt), 2) AS avg_cnt_reviews
          FROM (SELECT username, COUNT(review_id) AS cnt
                FROM reviews
                GROUP BY username
                HAVING username IN (SELECT username 
                                    FROM ratings
                                    GROUP BY username
                                    HAVING COUNT(rating_id) > 50)) query_in_1 
                
       ''' 

In [25]:
select(sql)

avg_cnt_reviews
0            24.33

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно 24.

Дополнительное исследование, для проверки теории

In [26]:
sql = ''' SELECT username, COUNT(review_id) AS cnt
                           FROM reviews
                           GROUP BY username
                           ORDER BY COUNT(review_id) desc
                           LIMIT 10
        '''                                          

In [27]:
select(sql)

username  cnt
0         susan85   29
1     sfitzgerald   28
2      martinadam   27
3          eallen   26
4     lesliegibbs   26
5         tanya01   26
6       richard89   26
7  jennifermiller   25
8     mariokeller   25
9           amy97   25

Те пользователи, что ставят много оценок, теже, чаще всего и обзоров публикуют больше. Положительная корреляция у этих двух признаков налицо.

### Общий вывод


- У нас таблицы с данными о книгах, об авторах, об издательствах, о пользовательских оценках и обзорах.  
  - Можно было заменить тип данных в соответствии с типом данных представленных в столбце дат  — в формат даты, но мы не стали этого делать.
  - Пропусков и дубликатов не обнаружили.

- Провели исследование:
  - После 2000 года издано 819 книг, информция о книгах (по названию) имеется в одном случае более чем в одной записи.
  - Посчитали  количество обзоров и среднюю оценку для каждой книги, по ним м.б. судить о том какие книги более интересны читателю.
  - Определили издательство, которое выпустило наибольшее число книг толще 50 страниц, и это Penguin Books - одно из самых популярных в мире.
  - Джоан Роулинг с Мэри Грандпре — британский писатель и американский иллюстратор и писатель, эти два человека наиболее известные своим авторством и своими иллюстрациями изданий книг о Гарри Поттере. И в нашем рейтинге они имеют смаую высокую среднюю оценку книг — учитывая только книги с 50 и более оценками.¶
  - Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно 24.

Из всего вышесказанного данное исследование поможет сформулировать ценностное предложение для сервиса чтения книг по подписке. Т.е. нужно ориентироваться на количество обзоров, оценок книг, а также на идательства (их популярность) и авторов. 
- PS Ценностное предложение - это ясное и простое изложение преимуществ, которые потребители получат при покупке продукта или услуги.